hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/YugabyteDBDialect.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.community.dialect;
+
+import org.hibernate.boot.model.TypeContributions;
+import org.hibernate.boot.registry.classloading.spi.ClassLoaderService;
+import org.hibernate.boot.registry.classloading.spi.ClassLoadingException;
+import org.hibernate.dialect.*;
+import org.hibernate.engine.jdbc.dialect.spi.DialectResolutionInfo;
+import org.hibernate.service.ServiceRegistry;
+import org.hibernate.type.JavaObjectType;
+import org.hibernate.type.descriptor.jdbc.*;
+import org.hibernate.type.descriptor.jdbc.spi.JdbcTypeRegistry;
+
+import java.sql.Types;
+
+public class YugabyteDBDialect extends PostgreSQLDialect {
```

Comment:
The `PostgreSQLDialect` will have to introduce a `getPostgreSQLVersion()` method similar to what we have done with `MySQLDialect#getMySQLVersion()` and override this new method in this class to report the "PostgreSQL compatible version" i.e. the version which matches upstream PostgreSQL the most.
All uses of `getVersion()` within `PostgreSQLDialect` and `PostgreSQLSqlAstTranslator` will have to change to `getPostgreSQLVersion()`. Look at how this is done in `MySQLDialect` and `MySQLSqlAstTranslator`.

Reply:
You can use the following patch. Alternatively you'll have to override `initializeFunctionRegistry` and `supportsFetchClause` to make sure the dialect behaves correctly based on the yugabyte version, since in the PostgreSQLDialect the mentioned methods are implemented in terms of the PostgreSQL version.

```
Index: hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLSqlAstTranslator.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLSqlAstTranslator.java b/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLSqlAstTranslator.java
--- a/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLSqlAstTranslator.java	(revision 22b42b1674563d351f87713ef4b4b806f37b8954)
+++ b/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLSqlAstTranslator.java	(date 1699542082658)
@@ -199,17 +199,17 @@
 
 	@Override
 	protected boolean supportsRecursiveSearchClause() {
-		return getDialect().getVersion().isSameOrAfter( 14 );
+		return getDialect().getPostgreSQLVersion().isSameOrAfter( 14 );
 	}
 
 	@Override
 	protected boolean supportsRecursiveCycleClause() {
-		return getDialect().getVersion().isSameOrAfter( 14 );
+		return getDialect().getPostgreSQLVersion().isSameOrAfter( 14 );
 	}
 
 	@Override
 	protected boolean supportsRecursiveCycleUsingClause() {
-		return getDialect().getVersion().isSameOrAfter( 14 );
+		return getDialect().getPostgreSQLVersion().isSameOrAfter( 14 );
 	}
 
 	@Override
@@ -276,4 +276,9 @@
 		arithmeticExpression.getRightHandOperand().accept( this );
 		appendSql( CLOSE_PARENTHESIS );
 	}
+
+	@Override
+	public PostgreSQLDialect getDialect() {
+		return (PostgreSQLDialect) super.getDialect();
+	}
 }
Index: hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java b/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java
--- a/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java	(revision 22b42b1674563d351f87713ef4b4b806f37b8954)
+++ b/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java	(date 1699542035124)
@@ -180,6 +180,10 @@
 		return MINIMUM_VERSION;
 	}
 
+	public DatabaseVersion getPostgreSQLVersion() {
+		return getVersion();
+	}
+
 	@Override
 	public boolean getDefaultNonContextualLobCreation() {
 		return true;
@@ -645,7 +649,7 @@
 		functionFactory.arrayRemoveIndex_unnest( true );
 		functionFactory.arraySlice_operator();
 		functionFactory.arrayReplace();
-		if ( getVersion().isSameOrAfter( 14 ) ) {
+		if ( getPostgreSQLVersion().isSameOrAfter( 14 ) ) {
 			functionFactory.arrayTrim_trim_array();
 		}
 		else {
@@ -1361,7 +1365,7 @@
 			case PERCENT_WITH_TIES:
 				return false;
 			case ROWS_WITH_TIES:
-				return getVersion().isSameOrAfter( 13 );
+				return getPostgreSQLVersion().isSameOrAfter( 13 );
 		}
 		return false;
 	}

```